src/org/jitsi/impl/neomedia/transform/RetransmissionRequesterDelegate.java

```diff
@@ -0,0 +1,582 @@
+     */
+    @Override
+    public RawPacket reverseTransform(RawPacket pkt)
+    {
+        Long ssrc;
+        int seq;
+
+        MediaFormat format = stream.getFormat(pkt.getPayloadType());
+        if (format == null)
+        {
+            ssrc = null;
+            seq = -1;
+
+            logger.warn("format_not_found" +
+                ",stream_hash=" + stream.hashCode());
+        }
+        else if (Constants.RTX.equalsIgnoreCase(format.getEncoding()))
+        {
+            MediaStreamTrackReceiver receiver
+                = stream.getMediaStreamTrackReceiver();
+
+            RTPEncodingDesc encoding = receiver.findRTPEncodingDesc(pkt);
+
+            if (encoding != null)
+            {
+                ssrc = encoding.getPrimarySSRC();
+                seq = pkt.getOriginalSequenceNumber();
+            }
+            else
+            {
+                ssrc = null;
+                seq = -1;
+
+                logger.warn("encoding_not_found" +
+                    ",stream_hash=" + stream.hashCode());
+            }
+        }
+        else
+        {
+            ssrc = pkt.getSSRCAsLong();
+            seq = pkt.getSequenceNumber();
+        }
+
+
+        if (ssrc != null)
+        {
+            // TODO(gp) Don't NACK higher temporal layers.
+            Requester requester = getOrCreateRequester(ssrc);
+            requester.received(seq);
+        }
+
+        return pkt;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getTimeUntilNextRun()
+    {
+        long now = timeProvider.getTime();
+        List<Requester> dueRequesters = getDueRequesters(now);
+        if (dueRequesters.isEmpty())
+        {
+            return WAKEUP_INTERVAL_MILLIS;
+        }
+        else
+        {
+            if (logger.isTraceEnabled())
+            {
+                Requester nextDueRequester = dueRequesters.get(0);
+                logger.trace(hashCode() + ": Next nack is scheduled for ssrc " +
+                    nextDueRequester.ssrc + " at " +
+                    Math.max(nextDueRequester.nextRequestAt, 0) +
+                    ".  (current time is " + now + ")");
+            }
+            return Math.max(dueRequesters.get(0).nextRequestAt - now, 0);
+        }
+    }
+
+    public void setWorkReadyCallback(Runnable workReadyCallback)
+    {
+        this.workReadyCallback = workReadyCallback;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void run()
+    {
+        long now = timeProvider.getTime();
+        if (logger.isTraceEnabled())
+        {
+            logger.trace(hashCode() + " running at " + now);
+        }
+        List<Requester> dueRequesters = getDueRequesters(now);
+        if (logger.isTraceEnabled())
+        {
+            logger.trace(hashCode() + " has " + dueRequesters.size() + " due requesters");
+        }
+        List<NACKPacket> nackPackets = createNackPackets(now, dueRequesters);
```

Comment:
Do we want to hold the lock between getDueRequesters and createNackPackets?

Reply:
I'm worried about a thread calling Requester.received() while we're creating nack packets

Reply:
yeah i think you're right.  and this means we'd need to take the lock at the call to `received` (currently in `reverseTransform`), is that something we regularly do in the packet path?  if not might need to look at another way to do it to avoid the locking.